hooks/lib/ecr-registry-provider.bash

```diff
@@ -0,0 +1,106 @@
+login() {
+  $(aws ecr get-login --no-include-email)
+}
+
+get_registry_url() {
+  local repository_name
+  repository_name="$(get_ecr_repository_name)"
+  aws ecr describe-repositories \
+    --repository-names "${repository_name}" \
+    --output text \
+    --query 'repositories[0].repositoryUri'
+}
+
+ecr_exists() {
+  local repository_name="${1}"
+  aws ecr describe-repositories \
+    --repository-names "${repository_name}" \
+    --output text \
+    --query 'repositories[0].registryId'
+}
+
+get_ecr_arn() {
+  local repository_name="${1}"
+  aws ecr describe-repositories \
+    --repository-names "${repository_name}" \
+    --output text \
+    --query 'repositories[0].repositoryArn'
+}
+
+get_ecr_tags() {
+local result=$(cat <<EOF
+{
+    "tags": []
+}
+EOF
+)
+  while IFS='=' read -r name _ ; do
+    if [[ $name =~ ^(BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ECR_TAGS_) ]] ; then
+      # Handle plain key=value, e.g
+      # ecr-tags:
+      #   KEY_NAME: 'key-value'
+      key_name=$(echo "${name}" | sed 's/^BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ECR_TAGS_//')
```

Comment:
reckon it's worth "standardising" on the interface, e.g. this would become `tags` rather than `ecr-tags`?

Reply:
we'd probably want both for backwards compat, but that's easy :-)

Reply:
Actually - I don't know whether GCR has an equivalent to this ECR concept. I'm unfamiliar with ECR in enough depth to know what to look for, to be honest!

Reply:
In that case, reckon it's worth splitting this into another plugin? e.g. `docker-gcr-cache-buildkite-plugin`? We can align on the interface fairly easily if people want to switch between them

I'm just wondering how easy future updates are going to be given we don't use GCR, and you don't use ECR :)

Reply:
The closest GCR has is the concept of labels, which get applied to the GCR itself (which actually turns into a user-visible GCS bucket). It's not possible to label images/image-families/channels within the GCR.

We do use ECR; we operate with AWS within the borders of China, where Google doesn't operate. We have a set of issues with traffic that needs to cross the Great Firewall of China So, I might not use ECR day to day, but I'd prefer to have a single plugin with multiple implementations within it. One of the things I'd like to be able to do is push to a number of registries, for example.

Reply:
Makes sense, good reasons :)